File: historyscan-back/src/main/java/io/sch/historyscan/domain/contexts/analysis/common/EnumIgnoredCodeBaseFiles.java
package io.sch.historyscan.domain.contexts.analysis.common;

import io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem.FilePath;

import java.util.Arrays;

import static io.sch.historyscan.domain.contexts.analysis.common.EnumPathType.FILE;
import static io.sch.historyscan.domain.contexts.analysis.common.EnumPathType.FOLDER;

public enum EnumIgnoredCodeBaseFiles {
    DEV_NULL("/dev/null", FILE),
    PACKAGE_LOCK_JSON("package-lock.json", FILE),
    YARN_LOCK("yarn.lock", FILE),
    LIBS("libs", FOLDER),
    LIB("lib", FOLDER),
    JS_MAP("js.map", FILE),
    MAP("map", FILE),
    NOD_MODULES("node_modules", FOLDER),
    DIST("dist", FOLDER),
    BUILD("build", FOLDER),
    CACHE(".cache", FILE),
    DOT_TEMP(".temp", FILE),
    TEMP("temp", FOLDER),
    COVERAGE("coverage", FOLDER),
    DOCS("docs", FOLDER),
    D_TS(".d.ts", FILE),
    RELEASE("release", FOLDER),
    CLASS(".class", FILE),
    JAR(".jar", FILE),
    BIN_FOLDER("bin", FOLDER),
    BIN_FILE("bin", FILE),
    TARGET("target", FOLDER),
    OUT("out", FOLDER),
    GENERATED("generated", FOLDER),
    GRADLEW("gradlew", FILE),
    DOT_GRADLE(".gradle", FOLDER),
    GRADLE("gradle", FOLDER),
    GITATTRIBUTES("gitattributes", FILE),
    DOCUMENTATION("documentation", FOLDER),
    PACKAGE_INFO_JAVA("package-info.java", FILE),
    IDEA_INTELLIJ(".idea", FOLDER),
    HTML_REPORT("htmlreport", FOLDER),
    RUN("run", FOLDER),
    ;

    private final String title;
    private final EnumPathType type;

    EnumIgnoredCodeBaseFiles(String title, EnumPathType type) {
        this.title = title;
        this.type = type;
    }

    public static boolean isIgnored(String codebasesFolder, String path, boolean isFile) {
        var commonPart = FilePath.commonPartOf(codebasesFolder, path);
        var filteredPath = path.substring(commonPart.length());
        return Arrays.stream(values()).anyMatch(ignoredFile -> isIgnored(filteredPath, isFile, ignoredFile));
    }


    private static boolean isIgnored(String path, boolean isFile, EnumIgnoredCodeBaseFiles ignoredFile) {
        if (path.isEmpty()) {
            return false;
        }
        final String formattedPath = path.replace("\\", "/").trim().toLowerCase();
        if (isFile) {
            return (ignoredFile.type == FOLDER && formattedPath.contains(ignoredFile.title + "/"))
                   || formattedPath.substring(formattedPath.lastIndexOf("/") + 1)
                              .equals(ignoredFile.title) && ignoredFile.type == FILE;
        } else {
            return formattedPath.contains(ignoredFile.title + "/")
                   || formattedPath.contains("/" + ignoredFile.title)
                   || ignoredFile.title.contains(formattedPath);
        }
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/domain/contexts/analysis/common/EnumIgnoredCodeBaseFiles.kt
package io.sch.historyscan.domain.contexts.analysis.common

import io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem.*
import java.util.*

enum class EnumIgnoredCodeBaseFiles(private val title: String, private val type: EnumPathType) {
    DEV_NULL("/dev/null", EnumPathType.FILE),
    PACKAGE_LOCK_JSON("package-lock.json", EnumPathType.FILE),
    YARN_LOCK("yarn.lock", EnumPathType.FILE),
    LIBS("libs", EnumPathType.FOLDER),
    LIB("lib", EnumPathType.FOLDER),
    JS_MAP("js.map", EnumPathType.FILE),
    MAP("map", EnumPathType.FILE),
    NOD_MODULES("node_modules", EnumPathType.FOLDER),
    DIST("dist", EnumPathType.FOLDER),
    BUILD("build", EnumPathType.FOLDER),
    CACHE(".cache", EnumPathType.FILE),
    DOT_TEMP(".temp", EnumPathType.FILE),
    TEMP("temp", EnumPathType.FOLDER),
    COVERAGE("coverage", EnumPathType.FOLDER),
    DOCS("docs", EnumPathType.FOLDER),
    D_TS(".d.ts", EnumPathType.FILE),
    RELEASE("release", EnumPathType.FOLDER),
    CLASS(".class", EnumPathType.FILE),
    JAR(".jar", EnumPathType.FILE),
    BIN_FOLDER("bin", EnumPathType.FOLDER),
    BIN_FILE("bin", EnumPathType.FILE),
    TARGET("target", EnumPathType.FOLDER),
    OUT("out", EnumPathType.FOLDER),
    GENERATED("generated", EnumPathType.FOLDER),
    GRADLEW("gradlew", EnumPathType.FILE),
    DOT_GRADLE(".gradle", EnumPathType.FOLDER),
    GRADLE("gradle", EnumPathType.FOLDER),
    GITATTRIBUTES("gitattributes", EnumPathType.FILE),
    DOCUMENTATION("documentation", EnumPathType.FOLDER),
    PACKAGE_INFO_JAVA("package-info.java", EnumPathType.FILE),
    IDEA_INTELLIJ(".idea", EnumPathType.FOLDER),
    HTML_REPORT("htmlreport", EnumPathType.FOLDER),
    RUN("run", EnumPathType.FOLDER),
    ;

    companion object {
        @JvmStatic
        fun isIgnored(codebasesFolder: String, path: String, isFile: Boolean): Boolean {
            val commonPart: String = FilePath.commonPartOf(codebasesFolder, path)
            val filteredPath = path.substring(commonPart.length)
            return Arrays.stream(entries.toTypedArray()).anyMatch { ignoredFile: EnumIgnoredCodeBaseFiles -> isIgnored(filteredPath, isFile, ignoredFile) }
        }

        private fun isIgnored(path: String, isFile: Boolean, ignoredFile: EnumIgnoredCodeBaseFiles): Boolean {
            if (path.isEmpty()) {
                return false
            }
            val formattedPath = path.replace("\\", "/").trim { it <= ' ' }.lowercase(Locale.getDefault())
            return if (isFile) {
                ((ignoredFile.type == EnumPathType.FOLDER && formattedPath.contains(ignoredFile.title + "/"))
                        || (formattedPath.substring(formattedPath.lastIndexOf("/") + 1)
                        == ignoredFile.title) && ignoredFile.type == EnumPathType.FILE)
            } else {
                (formattedPath.contains(ignoredFile.title + "/")
                        || formattedPath.contains("/" + ignoredFile.title)
                        || ignoredFile.title.contains(formattedPath))
            }
        }
    }
}
